sql: internal error: unable to vectorize execution plan: ... expected NULL:::RECORD to be of type tuple, found type unknown

roachtest.sqlsmith/setup=seed/setting=no-mutations [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4099953&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4099953&tab=artifacts#/sqlsmith/setup=seed/setting=no-mutations) on master @ [bbd8658d09681135c30e07650798768ca0063f09](https://github.com/cockroachdb/cockroach/commits/bbd8658d09681135c30e07650798768ca0063f09):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed/setting=no-mutations/run_1
	sqlsmith.go:259,sqlsmith.go:298,test_runner.go:780: error: pq: internal error: unable to vectorize execution plan: CASE WHEN "$9" THEN (B'0000000000', ARRAY[18078, 29977, -28375], ('4 mons 716 days 17:39:04.590025':::INTERVAL, (-32768):::INT8)) ELSE (B'1100011111', ARRAY[], NULL:::RECORD) END: incompatible value type: tuples [(B'0000000000', ARRAY[18078, 29977, -28375], ('4 mons 716 days 17:39:04.590025':::INTERVAL, (-32768):::INT8)) (B'1100011111', ARRAY[], NULL:::RECORD)] are not the same type: expected NULL:::RECORD to be of type tuple, found type unknown
		stmt:
		WITH
			with_67032 (col_400597)
				AS (
					SELECT
						CASE
						WHEN tab_163440._bool
						THEN (B'0000000000', ARRAY[18078:::INT8,29977:::INT8,(-28375):::INT8], ('4 mons 716 days 17:39:04.590025':::INTERVAL, (-32768):::INT8))
						ELSE (B'1100011111', ARRAY[]:::INT2[], NULL)
						END
							AS col_400597
					FROM
						defaultdb.public.seed@[0] AS tab_163440
				)
		SELECT
			cte_ref_19680.col_400597 AS col_400598
		FROM
			with_67032 AS cte_ref_19680
		WHERE
			true
		LIMIT
			15:::INT8;
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=seed/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-12233

roachtest.sqlsmith/setup=seed/setting=no-mutations [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4140354&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4140354&tab=artifacts#/sqlsmith/setup=seed/setting=no-mutations) on master @ [365b4da8bd02c06ee59d2130a56dec74ffc9ce21](https://github.com/cockroachdb/cockroach/commits/365b4da8bd02c06ee59d2130a56dec74ffc9ce21):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed/setting=no-mutations/run_1
	sqlsmith.go:259,sqlsmith.go:298,test_runner.go:780: error: pq: internal error: crdb_internal.replication_stream_progress(): job -1109278611987283860: log-job: UpdateDeadline() called on leaf txn
		stmt:
		SELECT
			crdb_internal.replication_stream_progress((-1109278611987283860):::INT8::INT8, tab_268._string::STRING)::BYTES
				AS col_644
		FROM
			defaultdb.public.seed@seed__int8__float8__date_idx AS tab_268,
			defaultdb.public.seed AS tab_269
			JOIN defaultdb.public.seed AS tab_270 ON (tab_269._jsonb) = (tab_270._jsonb)
		ORDER BY
			tab_268._inet, tab_268._float4 DESC;
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=seed/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

> roachtest.sqlsmith/setup=seed/setting=no-mutations [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4108360&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4108360&tab=artifacts#/sqlsmith/setup=seed/setting=no-mutations) on master @ [c3d71ac887844bef174abb6dab2a4e1ce9270ab7](https://github.com/cockroachdb/cockroach/commits/c3d71ac887844bef174abb6dab2a4e1ce9270ab7):
> 
> ```
> The test failed on branch=master, cloud=gce:
> test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed/setting=no-mutations/run_1
> 	sqlsmith.go:278,sqlsmith.go:298,test_runner.go:780: ping node 1: dial tcp 104.196.200.208:26257: connect: connection refused
> 		previous sql:
> 		SELECT
> 			900146632:::INT8 AS col_103280,
> 			'hello':::greeting AS col_103281,
> 			tab_42412._timestamp AS col_103282,
> 			timezone(crdb_internal.replication_stream_progress(5928079676136251041:::INT8::INT8, tab_42412._string::STRING)::STRING::STRING, tab_42412._timestamp::TIMESTAMP)::TIMESTAMPTZ
> 				AS col_103283,
> 			tab_42412._interval AS col_103284,
> 			true AS col_103285,
> 			(-3.528206506834196767E+32):::DECIMAL AS col_103286
> 		FROM
> 			defaultdb.public.seed@seed__int8__float8__date_idx AS tab_42412
> 		LIMIT
> 			71:::INT8;
> ```
> 
> Help
> [This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=seed/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)

From node 1:
```
panic: UpdateDeadline() called on leaf txn

goroutine 35656 [running]:
panic({0x4716940, 0xc001ba6c90})
	/usr/local/go/src/runtime/panic.go:1147 +0x3a8 fp=0xc00ac2e180 sp=0xc00ac2e0c0 pc=0x488688
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).UpdateDeadline(0x182f691, {0x8297908, 0xc001ba68a0}, {0x0, 0x4ac0c80, 0x0})
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:774 +0x313 fp=0xc00ac2e260 sp=0xc00ac2e180 pc=0x12e89d3
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*leasedDescriptors).maybeUpdateDeadline(0xc0013948b0, {0x8297908, 0xc001ba68a0}, {0x821b0f0, 0xc00ab9ae70}, {0x0, 0x0})
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/leased_descriptors.go:221 +0x191 fp=0xc00ac2e2f0 sp=0xc00ac2e260 pc=0x1854971
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*leasedDescriptors).getResult(0xc0013948a0, {0x8297908, 0xc001ba68a0}, {0x821b0f0, 0xc00ab9ae70}, 0x1, {0x82fe6e0, 0xc000b4ec80}, {0x0, 0x0})
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/leased_descriptors.go:187 +0x477 fp=0xc00ac2e400 sp=0xc00ac2e2f0 pc=0x1854737
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*leasedDescriptors).getByName(0xc0013948a0, {0x8297908, 0xc001ba68a0}, {0x821b0f0, 0xc00ab9ae70}, 0x1, 0x1d, {0xc00283ab61, 0x4})
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/leased_descriptors.go:108 +0x372 fp=0xc00ac2e4d0 sp=0xc00ac2e400 pc=0x1853e52
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).getByName(0xc001394888, {0x8297908, 0xc001ba68a0}, 0xc00ab9ae70, {0x8396430, 0xc000489950}, {0x838b010, 0xadfcde0}, {0xc00283ab61, 0x4}, ...)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/descriptor.go:199 +0x2ed fp=0xc00ac2e5c8 sp=0xc00ac2e4d0 pc=0x184f90d
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).getObjectByNameIgnoringRequiredAndType(0xc001394888, {0x8297908, 0xc001ba68a0}, 0x0, {0xc00283ab5a, 0x6}, {0x4b43a4a, 0x6}, {0xc00283ab61, 0x4}, ...)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/object.go:194 +0x358 fp=0xc00ac2e6a0 sp=0xc00ac2e5c8 pc=0x1856198
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).getObjectByName(0xc0021b2900, {0x8297908, 0xc001ba68a0}, 0xc0021b28e0, {0xc00283ab5a, 0xc002357860}, {0x4b43a4a, 0x0}, {0xc00283ab61, 0x4}, ...)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/object.go:64 +0x1b2 fp=0xc00ac2e860 sp=0xc00ac2e6a0 pc=0x18552b2
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).GetObjectDesc(0xba892f, {0x8297908, 0xc001ba68a0}, 0x54, {0xc00283ab5a, 0x8}, {0x4b43a4a, 0x0}, {0xc00283ab61, 0x4}, ...)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/object.go:39 +0x9f fp=0xc00ac2e960 sp=0xc00ac2e860 pc=0x1854fbf
github.com/cockroachdb/cockroach/pkg/sql.(*planner).LookupObject(0xbac6bb, {0x8297908, 0xc001ba68a0}, {{0x0, 0x0, 0x0, 0x0, 0x0, 0x0}, 0x0, ...}, ...)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/resolver.go:200 +0x5d6 fp=0xc00ac2eaa0 sp=0xc00ac2e960 pc=0x3080ad6
github.com/cockroachdb/cockroach/pkg/sql/catalog/resolver.ResolveExisting({0x8297908, 0xc001ba68a0}, 0xc008fdb840, {0x7faab2b5a788, 0xc001394c48}, {{0x1, 0x0, 0x0, 0x0, 0x0, ...}, ...}, ...)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/resolver/resolver.go:440 +0x33b fp=0xc00ac2ece0 sp=0xc00ac2eaa0 pc=0x1846a9b
github.com/cockroachdb/cockroach/pkg/sql/catalog/resolver.ResolveExistingObject({0x8297908, 0xc001ba68a0}, {0x8310498, 0xc001394c48}, 0xc008fdb840, {{0x1, 0x0, 0x0, 0x0, 0x0, ...}, ...})
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/resolver/resolver.go:195 +0x173 fp=0xc00ac2eec8 sp=0xc00ac2ece0 pc=0x1844973
github.com/cockroachdb/cockroach/pkg/sql/catalog/resolver.ResolveExistingTableObject({0x8297908, 0xc001ba68a0}, {0x8310498, 0xc001394c48}, 0xc001395458, {{0x1, 0x0, 0x0, 0x0, 0x0, ...}, ...})
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/resolver/resolver.go:123 +0x15d fp=0xc00ac2efb0 sp=0xc00ac2eec8 pc=0x1843e1d
github.com/cockroachdb/cockroach/pkg/sql.(*optCatalog).ResolveDataSource(0xc001395440, {0x8297908, 0xc001ba68a0}, {0x98, 0xfc}, 0xc0021b3258)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/opt_catalog.go:209 +0x20f fp=0xc00ac2f1e8 sp=0xc00ac2efb0 pc=0x302f1ef
github.com/cockroachdb/cockroach/pkg/sql/opt.(*Metadata).CheckDependencies(0xc0059e45a0, {0x8297908, 0xc001ba68a0}, {0x8337988, 0xc001395440})
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/metadata.go:293 +0xf7 fp=0xc00ac2f2b0 sp=0xc00ac2f1e8 pc=0x1cde237
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*Memo).IsStale(0xc0059e45a0, {0x8297908, 0xc001ba68a0}, 0xc001394f08, {0x8337988, 0xc001395440})
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/memo.go:322 +0x493 fp=0xc00ac2f2e8 sp=0xc00ac2f2b0 pc=0x1fd36f3
github.com/cockroachdb/cockroach/pkg/sql.(*planner).prepareUsingOptimizer(0xc001394c48, {0x8297908, 0xc001ba68a0})
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:102 +0x8af fp=0xc00ac2f440 sp=0xc00ac2f2e8 pc=0x30594af
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).populatePrepared(0xc001394e00, {0x8297908, 0xc001ba68a0}, 0xc00b82e5c0, {0xc00b82e5c0, 0x1, 0xc0021b3608}, 0xc001394c48)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go:287 +0x22c fp=0xc00ac2f558 sp=0xc00ac2f440 pc=0x2f774cc
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).prepare.func1({0x8297908, 0xc001ba68a0}, 0x10)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go:220 +0x40a fp=0xc00ac2f640 sp=0xc00ac2f558 pc=0x2f7702a
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).prepare(0xc001394600, {0x8297908, 0xc001ba68a0}, {{{0x82c7338, 0xc00a617720}, {0xc00b6fac00, 0x73}, 0x1, 0x1}, {0xc00b6fad80, ...}, ...}, ...)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go:226 +0x2cd fp=0xc00ac2f740 sp=0xc00ac2f640 pc=0x2f7692d
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).addPreparedStmt(0xc001394600, {0x8297908, 0xc001ba68a0}, {0x0, 0x0}, {{{0x82c7338, 0xc00a617720}, {0xc00b6fac00, 0x73}, 0x1, ...}, ...}, ...)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go:111 +0x110 fp=0xc00ac2f840 sp=0xc00ac2f740 pc=0x2f76470
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execPrepare(0xc001394600, {0x8297908, 0xc001ba68a0}, {{0x0, 0x0}, {{0x82c7338, 0xc00a617720}, {0xc00b6fac00, 0x73}, 0x1, ...}, ...})
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go:53 +0x23b fp=0xc00ac2f9f8 sp=0xc00ac2f840 pc=0x2f760bb
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc001394600, {0x8297860, 0xc008fdb7c0})
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1844 +0x1226 fp=0xc00ac2fe88 sp=0xc00ac2f9f8 pc=0x2f5dbe6
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc001394600, {0x82b39b0, 0xc0025db3a0}, 0xf0f00c00194e7d0, {0x1027380, 0xc0025da300, 0x0}, 0xc000b43830)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1676 +0x285 fp=0xc00ac2ff40 sp=0xc00ac2fe88 pc=0x2f5c7c5
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx.func1()
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:205 +0x8d fp=0xc00ac2ffe0 sp=0xc00ac2ff40 pc=0x30228ed
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1581 +0x1 fp=0xc00ac2ffe8 sp=0xc00ac2ffe0 pc=0x4bc501
created by github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:204 +0x5f1
```

Duplicate of #75121

--- a/pkg/sql/logictest/testdata/logic_test/tuple
+++ b/pkg/sql/logictest/testdata/logic_test/tuple
@@ -1173,3 +1173,15 @@ FROM
       (NULL, (2:::DECIMAL, ARRAY[(1::INT8, 2::INT8, 3::INT8)]))
   )
     AS t2 (c3, c4)
+
+# Regression test #74729. Correctly handle NULLs annotated as a tuple type.
+subtest regression_74729
+
+statement ok
+SELECT CASE WHEN true THEN ('a', 2) ELSE NULL:::RECORD END
+
+statement ok
+CREATE TABLE t74729 AS SELECT g % 2 = 1 AS _bool FROM generate_series(1, 5) AS g
+
+statement ok
+SELECT CASE WHEN _bool THEN (1, ('a', 2)) ELSE (3, NULL) END FROM t74729

--- a/pkg/sql/sem/tree/type_check.go
+++ b/pkg/sql/sem/tree/type_check.go
@@ -2570,7 +2570,10 @@ func typeCheckSameTypedTupleExprs(
 func checkAllExprsAreTuplesOrNulls(ctx context.Context, semaCtx *SemaContext, exprs []Expr) error {
 	for _, expr := range exprs {
 		_, isTuple := expr.(*Tuple)
-		isNull := expr == DNull
+		isNull, err := isNullOrAnnotatedNullTuple(ctx, semaCtx, expr)
+		if err != nil {
+			return err
+		}
 		if !(isTuple || isNull) {
 			// We avoid calling TypeCheck on Tuple exprs since that causes the
 			// types to be resolved, which we only want to do later in type-checking.
@@ -2605,6 +2608,28 @@ func checkTupleHasLength(t *Tuple, expectedLen int) error {
 		return pgerror.Newf(pgcode.DatatypeMismatch, "expected tuple %v to have a length of %d", t, expectedLen)
 	}
 	return nil
+}
+
+// isNullOrAnnotatedNullTuple returns true if the given expression is a DNull or
+// a DNull that is wrapped by an AnnotateTypeExpr with a type identical to
+// AnyTuple (e.g., NULL:::RECORD). An AnnotateTypeExpr should never have a tuple
+// type not identical to AnyTuple, because other tuple types could only be
+// expressed as composite types, and these are not yet supported. If the
+// AnnotateTypeExpr has a non-tuple type, the function returns false.
+func isNullOrAnnotatedNullTuple(
+	ctx context.Context, semaCtx *SemaContext, expr Expr,
+) (bool, error) {
+	if expr == DNull {
+		return true, nil
+	}
+	if annotate, ok := expr.(*AnnotateTypeExpr); ok && annotate.Expr == DNull {
+		annotateType, err := ResolveType(ctx, annotate.Type, semaCtx.GetTypeResolver())
+		if err != nil {
+			return false, err
+		}
+		return annotateType.Identical(types.AnyTuple), nil
+	}
+	return false, nil
 }
 
 type placeholderAnnotationVisitor struct {

--- a/pkg/sql/sem/tree/type_check_test.go
+++ b/pkg/sql/sem/tree/type_check_test.go
@@ -161,6 +161,12 @@ func TestTypeCheck(t *testing.T) {
 		{`(1:::INT8, 2:::INT8)`, `(1:::INT8, 2:::INT8)`},
 		{`(ROW (1,2))`, `(1:::INT8, 2:::INT8)`},
 		{`ROW(1:::INT8, 2:::INT8)`, `(1:::INT8, 2:::INT8)`},
+		// Regression test #74729. Correctly handle NULLs annotated as a tuple
+		// type.
+		{
+			`CASE WHEN true THEN ('a', 2) ELSE NULL:::RECORD END`,
+			`CASE WHEN true THEN ('a':::STRING, 2:::INT8) ELSE NULL END`,
+		},
 
 		{`((ROW (1) AS a)).a`, `1:::INT8`},
 		{`((('1', 2) AS a, b)).a`, `'1':::STRING`},

